subsys/net/lib/azure_fota/azure_fota.c

```diff
@@ -0,0 +1,277 @@
+
+	/* The properties.desired.firmware object has the excpected structure:
+	 *	properties.desired.firmware : {
+	 *		fwVersion : <version>,
+	 *		fwLocation : {
+	 *			host : <hostname>,
+	 *			path : <file path>
+	 *		},
+	 *		fwCheckValue : <value>
+	 *	}
+	 */
+
+	root_obj = cJSON_Parse(msg);
+	if (root_obj == NULL) {
+		printk("Could not parse message as JSON\n");
+		return -ENOMSG;
+	}
+
+	/* Desired object */
+	desired_obj = cJSON_GetObjectItemCaseSensitive(root_obj, "desired");
+	if (desired_obj == NULL) {
+		/* The received object can be the 'desired' object only, or
+		 * the whole device twin properties object. If there is no
+		 * 'desired' object to be found, assume it's the root object.
+		 */
+		LOG_DBG("No 'desired' object found, assuming it's root object");
+		desired_obj = root_obj;
+	}
+
+	/* Firmware update information object */
+	fw_obj = cJSON_GetObjectItemCaseSensitive(desired_obj, "firmware");
+	if (fw_obj == NULL) {
+		LOG_ERR("Failed to parse 'firmware' object");
+		goto clean_exit;
+	}
+
+	/* Firmware version */
+	fw_version_obj = cJSON_GetObjectItemCaseSensitive(fw_obj, "fwVersion");
+	if (fw_version_obj == NULL) {
+		printk("No fwVersion object found");
+		goto clean_exit;
+	}
+
+	if (cJSON_IsString(fw_version_obj)) {
+		fw_version = atoi(fw_version_obj->valuestring);
+	} else if (cJSON_IsNumber(fw_version_obj)) {
+		fw_version = fw_version_obj->valueint;
+	} else {
+		printk("Invalid fwVersion format received\n");
+		goto clean_exit;
+	}
+
+	LOG_DBG("Firmware update received for version %d", fw_version);
+
+	/* Firmware image location */
+	fw_location_obj = cJSON_GetObjectItemCaseSensitive(fw_obj,
+							   "fwLocation");
+	if (fw_location_obj == NULL) {
+		printk("No fwLocation object found");
+		goto clean_exit;
+	}
+
+	/* Firmware image hostname */
+	fw_hostname_obj = cJSON_GetObjectItemCaseSensitive(fw_location_obj,
+							   "host");
+	if (fw_hostname_obj == NULL) {
+		printk("No hostname object found");
+		goto clean_exit;
+	}
+
+	if (strlen(fw_hostname_obj->valuestring) >= sizeof(hostname)) {
+		LOG_ERR("Received hostname larger (%d) than buffer (%d)",
+			strlen(fw_hostname_obj->valuestring), sizeof(hostname));
+		goto clean_exit;
+	}
+
+	memcpy(hostname, fw_hostname_obj->valuestring, sizeof(hostname));
+
+	/* Firmware image file path */
+	fw_path_obj = cJSON_GetObjectItemCaseSensitive(fw_location_obj, "path");
+	if (fw_path_obj == NULL) {
+		printk("No path object found");
+		goto clean_exit;
+	}
+
+	if (strlen(fw_path_obj->valuestring) >= sizeof(file_path)) {
+		LOG_ERR("Received file path larger (%d) than buffer (%d)",
+			strlen(fw_path_obj->valuestring), sizeof(file_path));
+		goto clean_exit;
+	}
+
+	memcpy(file_path, fw_path_obj->valuestring, sizeof(file_path));
+
+	fw_check_obj = cJSON_GetObjectItemCaseSensitive(fw_obj, "fwCheckValue");
+	if (fw_check_obj == NULL) {
+		printk("Failed to parse 'fwCheckValue' object");
+		goto clean_exit;
+	}
+
+	/* TODO: Do something sensible with this check value */
```

Comment:
The [official guide](https://docs.microsoft.com/en-us/azure/iot-hub/tutorial-firmware-update) does not provide any recommendation of what to do with the `fwCheckValue` field.

Looking at [some sample IoT clients](https://github.com/search?q=org%3AAzure-Samples+fwPackageCheckValue&type=Code) this should contain the checksum of the firmware file which the device can use to verify that it downloaded the file correctly.

We don't do this [in the AWS FOTA](https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/include/net/aws_fota.html) either because the integrity check is part of the FW file itself and the modem handles that.

So I recommend to drop it.

Reply:
Yeah, sounds reasonable to drop that field, as you say, it will be checked anyway by the respective bootloader receiving the image.